### ISSUE TITLE
Memory Leak reloading file??

### ISSUE THREAD

**Issue by JPHammonds** [prjemian/spec2nexus#127]
I am working on reloading spec files so that the user can reload to update the data as it is collected.  I have been looking at memory use as I load the data over and over again.  It seems that the memory used when a spec file is open is not reclaimed as the file is read again. 

Presently I am basically using 

self.specDataFile = None
self.specDataFile = SpecDataFile(fileName)

Each time I read the data the amount of memory shown by the Activity Monitor (Mac) increases by a couple megabytes.  This is just loading the file and then loading again (little interaction with the rest of the data.  I have even reproduced this in a small python program that does

>>> from spec2nexus.spec import SpecDataFile
>>> import gc
>>> a = SpecDataFile("/Users/hammonds/RSM_DATA/DanielHaskel/Brian-Nick/Fluorescence/lineup")
>>> a = None
>>> gc.collect()

I through in the gc stuff to try to force the release.  Even if this does not release right away, I would assume that I would start to see some of this memory being reused.

I have even played with adding some __del__ methods for the classes in spec.py and get similar results.  

Up till now did not worry about this but if the user is going to reload many times as data is collected this could become more of an issue.

**Comment by prjemian** [comment]
There must be additional reference(s) to one or more objects create in `SpecDataFile()`.

possible help to diagnose:

tool | link 
----- | -----
search |  https://duckduckgo.com/?q=diagnose+memory+leak+in++python+code&t=ffsb&ia=web
`gc`  |    https://stackoverflow.com/questions/17624114/how-to-diagnose-potential-memory-leak-in-python-program
`resource`  |   https://chase-seibert.github.io/blog/2013/08/03/diagnosing-memory-leaks-python.html
`tracemalloc`  |   https://blog.fugue.co/2017-03-06-diagnosing-and-fixing-memory-leaks-in-python.html


## Basically:

```
import gc

#  To debug a leaking program call:
gc.set_debug(gc.DEBUG_LEAK)


# the amount of memory your process is using :

import resource
print 'Memory usage: %s (kb)' % resource.getrusage(resource.RUSAGE_SELF).ru_maxrss

```

**Comment by prjemian** [comment]
Work this up in a unit test module.

**Comment by prjemian** [comment]
Not seeing a memory leak with this test program:

<details>
<summary>program</summary>
<p>

```
#!/usr/bin/env python

"""
try to provoke the memory leak described

https://github.com/prjemian/spec2nexus/issues/127

file: /Users/hammonds/RSM_DATA/DanielHaskel/Brian-Nick/Fluorescence/lineup
We don't have that one.
"""


import gc
import os
import resource
from spec2nexus.spec import SpecDataFile
import sys
import time

_path = os.path.dirname(__file__)
# TESTFILE = "tests/data/issue119_data.txt"
TESTFILE = os.path.join(_path, "data", "issue119_data.txt")


def get_memory():
    return resource.getrusage(resource.RUSAGE_SELF).ru_maxrss

def task():
    gc.collect()
    before = get_memory()
    # gc.set_debug(gc.DEBUG_LEAK)
    a = SpecDataFile(TESTFILE)
    after = get_memory()
    a = None
    gc.collect()
    final = get_memory()

    print(f"{before}  {after}  {final}")


if __name__ == "__main__":
    for i in range(65):
        task()
        time.sleep(1.0)
```

</p>
</details>

<details>
<summary>output</summary>
<p>

```
(base) mintadmin@mint-vm:~/.../eclipse/spec2nexus$ /home/mintadmin/Apps/anaconda/bin/python /home/mintadmin/Documents/eclipse/spec2nexus/tests/leak_check.py
10248  57900  57964
57964  58504  58504
58504  58504  58504
58504  58504  58504
58504  58504  58504
58504  58504  58504
58504  58504  58504
58504  58504  58504
58504  58504  58504
58504  58504  58504
58504  58504  58504
58504  58504  58504
58504  58504  58504
58504  58504  58504
58504  58520  58520
58520  58520  58520
58520  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
58524  58524  58524
```

</p>
</details>

Quickly (two passes) reaches memory usage of ca. 58.5 MB and remains steady there.

@JPHammonds How can I permute this test to reproduce the leak?  Can you provide the data file in the question?  Looks like it is on a Mac file system.

**Comment by prjemian** [comment]
adjusted the reporting and number of iterations:

```
(base) mintadmin@mint-vm:~/.../eclipse/spec2nexus$ /home/mintadmin/Apps/anaconda/bin/python /home/mintadmin/Documents/eclipse/spec2nexus/tests/leak_check.py
{'time': 1574353778.5132792, 'before': 9.90234375, 'after': 56.4765625, 'final': 56.4765625}
{'time': 1574353779.5455828, 'before': 56.4765625, 'after': 56.8203125, 'final': 56.8203125}
{'time': 1574353780.5986793, 'before': 56.8203125, 'after': 56.8203125, 'final': 56.8203125}
{'time': 1574353781.6301248, 'before': 56.8203125, 'after': 56.8203125, 'final': 56.8203125}
{'time': 1574353782.6783078, 'before': 56.8203125, 'after': 56.8203125, 'final': 56.8203125}
{'time': 1574353783.722985, 'before': 56.8203125, 'after': 56.8203125, 'final': 56.8203125}
{'time': 1574353784.7548738, 'before': 56.8203125, 'after': 56.8203125, 'final': 56.8203125}
{'time': 1574353785.791578, 'before': 56.8203125, 'after': 56.8203125, 'final': 56.8203125}
{'time': 1574353786.8364105, 'before': 56.8203125, 'after': 56.8203125, 'final': 56.8203125}
```

**Comment by prjemian** [comment]
same conclusion, so far

**Comment by prjemian** [comment]
This article might also be useful: http://tech.labs.oliverwyman.com/blog/2008/11/14/tracing-python-memory-leaks/

**Comment by prjemian** [comment]
Running continuously on a different workstation, no leak observed after 25+ minutes.  Not that *time* is elapsed time, s, since program started.

started:

```
(base) bash-4.2$ python ./leak_check.py | tee /tmp/leak-check.txt
{'time': 0.2437911033630371, 'before': 7.4296875, 'after': 41.91015625, 'final': 41.9296875}
{'time': 1.2655704021453857, 'before': 41.9296875, 'after': 42.3515625, 'final': 42.3515625}
{'time': 2.287858724594116, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 3.3101346492767334, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 4.333557367324829, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 5.356088638305664, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 6.380166053771973, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 7.403045415878296, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 8.424983978271484, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 9.448151350021362, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 10.47160267829895, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 11.495447158813477, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 12.518433332443237, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 13.536943197250366, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
```

later

```
{'time': 1555.536453485489, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 1556.5569906234741, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 1557.5803134441376, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 1558.6309714317322, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 1559.6743824481964, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 1560.7139732837677, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 1561.7694942951202, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 1562.8171799182892, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
```

**Comment by prjemian** [comment]
@JPHammonds - Could your code be holding references to objects from previous calls to SpecDataFile()?  This code is just not leakign now.

```
{'time': 1746.9012978076935, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
```

**Comment by prjemian** [comment]
after a couple hours, still no evidence of memory leak now:

```
{'time': 7274.881997346878, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 7275.903183460236, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 7276.92507815361, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
{'time': 7277.9476935863495, 'before': 42.3515625, 'after': 42.3515625, 'final': 42.3515625}
```

**Comment by prjemian** [comment]
@JPHammonds - re-open or open new issue if memory leak suspected again

marked *wontfix* -- find a problem to be fixed

**Comment by JPHammonds** [comment]
Will do.
